### ISSUE TITLE
Large Docker Images Cause Authorization Errors

### ISSUE THREAD

**Issue by ryanvolum** [docker/hub-feedback#1944]
## Problem description
Issue #645 addresses an `unauthorized: authentication required` error that can appear when pushing a container image into dockerhub. There seem to be a number of culprits that cause the issue, many of which are explained/resolved in the comments. There remain several developers ([me included](https://github.com/docker/hub-feedback/issues/645#issuecomment-578982795)) who continue to get the issue in a specific case: when container images are very large. 

I'm able to successfully push a small image: 

```bash
C:\git\openai\test>docker push ryanvolum/py-foo
The push refers to repository [docker.io/ryanvolum/py-foo]
96ddd4466a34: Mounted from ryanvolum/py-test
aebb9c3ad291: Mounted from ryanvolum/py-test
8bdbbc38d97a: Mounted from ryanvolum/py-test
6e576898edb7: Mounted from ryanvolum/py-test
b8f75ed65728: Mounted from ryanvolum/py-test
2acb7acd7288: Mounted from ryanvolum/py-test
9e34e92e6b95: Mounted from ryanvolum/py-test
e24139c523b9: Mounted from ryanvolum/py-test
51997f5d204b: Mounted from ryanvolum/py-test
5c4b8cd3b09b: Mounted from ryanvolum/py-test
e44ff087319e: Mounted from ryanvolum/py-test
808fd332a58a: Mounted from ryanvolum/py-test
b16af11cbf29: Mounted from ryanvolum/py-test
37b9a4b22186: Mounted from ryanvolum/py-test
e0b3afb09dc3: Mounted from ryanvolum/py-test
6c01b5a53aac: Mounted from ryanvolum/py-test
2c6ac8e5063e: Mounted from ryanvolum/py-test
cc967c529ced: Mounted from ryanvolum/py-test
latest: digest: sha256:7d09978c5ebf7407ca38aae731c08933c9a1c5eeecc5c2ce61a33dc6e79d9548 size: 4093
```
When I then try a large (7GB image), the model takes substantial time to get "pushed" before throwing an error: 

```cmd
C:\git\openai\gpt-2-kubernetes>docker push ryanvolum/py-test
The push refers to repository [docker.io/ryanvolum/py-test]
7247f72e164b: Pushing [==================================================>]  6.234GB
7e8cdc06aab9: Pushed
cfcacc8b042c: Pushed
9988c5324564: Pushed
d6c9258470b2: Pushed
ac5349f9af57: Mounted from ryanvolum/gpt-2
47a79b5a2369: Mounted from ryanvolum/gpt-2
42958de951d6: Pushed
6effad8e07e5: Mounted from ryanvolum/gpt-2
62cbb606ac4a: Pushed
8c93aa82e26a: Pushed
4b9202833b21: Pushed
20048f8d31de: Pushed
c890837d995f: Mounted from ryanvolum/gpt-2
fcd654f1dc27: Pushed
becec4c05380: Pushed
8d267010480f: Mounted from ryanvolum/gpt-2
270f934787ed: Mounted from ryanvolum/gpt-2
02571d034293: Pushed
unauthorized: authentication required
```

Is there any chance that the time it takes to build the model is causing some kind of conflict between the host and dockerhub? I ask because several of the resolutions on #645 were related to the host's time. 

Steps to repro:
1. Create a large container image (mine is about 7GB)
2. Attempt to push the container image over a reasonably fast network. For me, this takes about ~15 minutes
3. Observe authorization error

**Comment by manishtomar** [comment]
The tokens provided by hub registry are 5 mins long. When a blob being pushed is too long I suspect the engine is trying to reuse a token that is expired. @thaJeztah Would you know if my suspicion is correct?

**Comment by thaJeztah** [comment]
Yes, I think that's the issue (I think there's an older issue about this); however (IIUC), the token should be only needed when _starting_ to push a blob (and a new token would be created for each blob to push); I'm not very familiar with that code, but @dmcgowan may know

**Comment by ryanvolum** [comment]
fwiw, I just switched over to my own private registry and am not blocked. I imagine this is a blocking issue for others though

**Comment by kikacu** [comment]
Does someone succeeded in pushing ~5GB image ... ?

**Comment by saborrie** [comment]
I found that I had this issue while trying to push an image on a slow broadband connection. It was only 200MB - not quite the 5GB above which suggests this is definitely a timeout problem and not a size problem.

I managed to get around it by setting up a remote docker host on an Azure virtual machine and configuring my local docker context to use a remote host:

https://docs.docker.com/engine/context/working-with-contexts/
```sh
docker context create my-context --docker host=ssh://myuser@myazurehostname.com
docker context use my-context

# now build and push to registry from the azure vm with better internet than my home wifi
```

I don't know if the higher internet speeds from an Azure vm will be enough to handle 5-7GB but hopefully this was enough to help someone who finds this post at some point.

**Comment by vdamery** [comment]
Are you still experiencing this issue ?

I will close it for now due to inactivity.

Feel free to reopen it if needed.
